File: apps/web/src/lib/orchestrator/schemas.ts
/**
 * Mirror of `apps/orchestrator/schemas.py` (Phase 6).
 *
 * Why hand-written and not codegen'd:
 *  - Phase 7 keeps tooling minimal (P10). Adding an OpenAPI codegen
 *    step would pull a build-time dependency on the FastAPI process.
 *  - The orchestrator surface is small and the test suite already
 *    pins the wire format on the Python side. Drift will surface as
 *    Vitest store failures and Playwright E2E failures, both of
 *    which run against the live FastAPI server.
 *
 * If this file diverges from `apps/orchestrator/schemas.py`, treat the
 * Python schemas as the source of truth and update here.
 */

export type SessionState =
  | "created"
  | "planning"
  | "awaiting_user_input"
  | "awaiting_approval"
  | "running"
  | "completed"
  | "failed";

export type ApprovalDecision = "approved" | "rejected";

export type ApprovalStatus = "pending" | "approved" | "rejected";

export type ClarificationStatus =
  | "pending"
  | "answered"
  | "cancelled"
  | "expired";

export type ClarificationAnswerType = "text" | "number" | "choice";

export type IntakeChoiceAction =
  | "use_existing_runtime_asset"
  | "review_existing_draft"
  | "adapt_similar_candidate"
  | "create_new_draft"
  | "ask_for_more_info";

export type IntakeNextStepStatus =
  | "runtime_asset_selected"
  | "draft_review_handoff"
  | "similar_candidate_review"
  | "draft_authoring_started"
  | "information_needed";

export type SessionEventKind =
  | "state_changed"
  | "natural_language_route_selected"
  | "intake_choice_recorded"
  | "intake_next_step_recorded"
  | "planning_completed"
  | "subtask_started"
  | "subtask_progress"
  | "subtask_completed"
  | "subtask_failed"
  | "approval_requested"
  | "approval_resolved"
  | "clarification_requested"
  | "clarification_resolved"
  | "error";

export interface SessionCreateRequest {
  client_label?: string | null;
}

export interface SessionCreateResponse {
  session_id: string;
  trace_id: string;
  state: SessionState;
  created_at: string;
}

export interface ChatRequest {
  session_id: string;
  message: string;
}

export interface SubtaskView {
  id: string;
  kind: "mechanical" | "organic";
  description: string;
  depends_on: readonly string[];
}

export interface LLMAssistShadowView {
  mode: string;
  succeeded: boolean;
  used_for_decision: boolean;
  model_name: string;
  provider_name: string;
  prompt_hash: string;
  schema_version: string;
  deterministic_route: string;
  confidence?: number | null;
  fallback_reason?: string | null;
  candidate_route?: string | null;
  candidate_subjects: readonly string[];
  candidate_style?: string | null;
  clarification_question?: string | null;
  requirement_candidate?: Record<string, unknown> | null;
  search_expansion?: Record<string, unknown> | null;
  raw_response_excerpt?: string | null;
}

export interface NaturalLanguageRouteView {
  selected_route: string;
  route_snapshot_id: string;
  reason: string;
  request_id?: string | null;
  runtime_asset_id?: string | null;
  draft_asset_id?: string | null;
  candidate_asset_ids: readonly string[];
  clarification_required: boolean;
  clarification_questions: readonly string[];
  generation_allowed: boolean;
  requirement: Record<string, unknown>;
  intake_decision?: AssetIntakeDecisionView | null;
  shadow_result?: LLMAssistShadowView | null;
}

export interface AssetIntakeDecisionView {
  status: string;
  reason: string;
  runtime_asset_id?: string | null;
  draft_asset_id?: string | null;
  candidate_asset_ids: readonly string[];
  reference_candidate_ids: readonly string[];
  new_draft_allowed: boolean;
  metadata: Record<string, unknown>;
}

export interface PendingGateView {
  gate_id: string;
  prompt: string;
  requested_at: string;
  status: ApprovalStatus;
}

export interface ClarificationQuestion {
  id: string;
  field: string;
  question: string;
  required: boolean;
  answer_type: ClarificationAnswerType;
  choices: readonly string[];
  unit?: string | null;
}

export interface PendingClarificationView {
  clarification_id: string;
  session_id: string;
  subtask_id?: string | null;
  trace_id?: string | null;
  reason: string;
  questions: readonly ClarificationQuestion[];
  source_decision: Record<string, unknown>;
  plan_snapshot_id?: string | null;
  requirement_snapshot_id?: string | null;
  requested_at: string;
  status: ClarificationStatus;
}

/**
 * Mirror of the backend IP/abuse pre-filter verdict (fallback ADR item 6).
 *
 * Typed with its four named fields rather than `Record<string, unknown>`
 * because a consumer reads `user_message_ko` to decide what the user is told;
 * an index signature lets a typo compile and surface only as a blank refusal
 * at runtime.
 */
export interface PrefilterDecisionView {
  decision: string;
  matched_rule_id: string | null;
  reason_code: string;
  user_message_ko: string;
}

export interface ChatResponse {
  session_id: string;
  trace_id: string;
  state: SessionState;
  subtasks: readonly SubtaskView[];
  clarifying_questions: readonly string[];
  pending_gate: PendingGateView | null;
  pending_clarification?: PendingClarificationView | null;
  route_result?: NaturalLanguageRouteView | null;
  /**
   * Present only when the request was refused before the planner ran. Every
   * planner-output field is empty in that reply, so a client that ignores this
   * field reports a safety refusal as a planner failure.
   */
  prefilter?: PrefilterDecisionView | null;
}

export interface ApproveRequest {
  session_id: string;
  gate_id: string;
  decision: ApprovalDecision;
  comments?: string | null;
}

/**
 * Mirror of `ApproveJobRequest` (Python). Body for `POST /approve/{job_id}`.
 * `gate_id` is carried in the URL path — backend uses `extra="forbid"` so
 * including it in the body is a 422.
 */
export interface ApproveJobRequest {
  session_id: string;
  decision: ApprovalDecision;
  comments?: string | null;
}

export interface ApproveResponse {
  session_id: string;
  gate_id: string;
  state: SessionState;
  decision: ApprovalDecision;
}

export type ClarificationAnswerValue = string | number;

export interface ClarifyRequest {
  session_id: string;
  clarification_id: string;
  answers: Record<string, ClarificationAnswerValue>;
}

export interface ClarifyResponse {
  session_id: string;
  clarification_id: string;
  accepted: boolean;
  state: SessionState;
  route_result?: NaturalLanguageRouteView | null;
  pending_clarification?: PendingClarificationView | null;
}

export interface IntakeChoiceRequest {
  session_id: string;
  action: IntakeChoiceAction;
  route_snapshot_id: string;
}

export interface IntakeChoiceResponse {
  session_id: string;
  accepted: boolean;
  action: IntakeChoiceAction;
  state: SessionState;
  route_snapshot_id: string;
  intake_status?: string | null;
  selected_route?: string | null;
  runtime_asset_id?: string | null;
  draft_asset_id?: string | null;
  candidate_asset_ids: readonly string[];
  reference_candidate_ids: readonly string[];
  draft_authoring_requested: boolean;
  runtime_execution_started: boolean;
  runtime_catalog_registered: boolean;
  product_ready: boolean;
  release_allowed: boolean;
  message: string;
}

export interface IntakeNextStepRequest {
  session_id: string;
  route_snapshot_id: string;
}

export interface IntakeNextStepView {
  session_id: string;
  action: IntakeChoiceAction;
  status: IntakeNextStepStatus;
  state: SessionState;
  route_snapshot_id: string;
  selected_route?: string | null;
  runtime_asset_id?: string | null;
  draft_asset_id?: string | null;
  candidate_asset_ids: readonly string[];
  reference_candidate_ids: readonly string[];
  draft_asset?: Record<string, unknown> | null;
  draft_authoring_status?: string | null;
  draft_authoring_reason?: string | null;
  // fallback ADR item 6: structured IP/abuse pre-filter decision when the capable
  // model was blocked before being invoked (optional additive field).
  prefilter?: Record<string, unknown> | null;
  review_required: boolean;
  runtime_execution_started: boolean;
  runtime_catalog_registered: boolean;
  product_ready: boolean;
  release_allowed: boolean;
  message: string;
}

export interface RuntimeAssetExecutionRequest {
  session_id: string;
  route_snapshot_id: string;
}

export interface RuntimeAssetExecutionResponse {
  session_id: string;
  state: SessionState;
  route_snapshot_id: string;
  runtime_asset_id: string;
  subtask_id: string;
  runtime_execution_started: boolean;
  runtime_catalog_registered: boolean;
  product_ready: boolean;
  release_allowed: boolean;
  next_step: IntakeNextStepView;
  manifest: SubtaskArtifactManifest;
  message: string;
}

export interface SessionEvent {
  session_id: string;
  trace_id: string;
  kind: SessionEventKind;
  state: SessionState;
  occurred_at: string;
  payload: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Artifact manifest (manifest ADR). Mirrors `modules.artifacts.schemas`.
// Only `subtask_completed` (and `subtask_failed`) events carry a manifest.
// ---------------------------------------------------------------------------

export type ArtifactKind =
  | "mechanical_mesh"
  | "organic_mesh"
  | "merged_mesh"
  | "gcode"
  | "threemf"
  | "validation_report"
  | "preview_mesh"
  | "preview_image";

export type ArtifactStatus = "ok" | "partial" | "quarantined";

export type ManifestStatus = "success" | "failed" | "partial" | "quarantined";

export interface ManifestError {
  stage: string;
  error_type: string;
  detail: string;
  metadata?: Record<string, unknown>;
}

export interface ArtifactRef {
  kind: ArtifactKind;
  relative_uri: string;
  mime: string;
  size_bytes: number;
  sha256: string;
  producer_adapter: string;
  producer_version: string;
  created_at: string;
  status: ArtifactStatus;
  metadata?: Record<string, unknown>;
}

export interface SubtaskArtifactManifest {
  schema_version: string;
  session_id: string;
  subtask_id: string;
  created_at: string;
  artifacts: readonly ArtifactRef[];
  status: ManifestStatus;
  errors: readonly ManifestError[];
  plan_snapshot_id?: string | null;
  approval_gate_id?: string | null;
  trace_id?: string | null;
}

export interface SubtaskCompletedPayload {
  subtask_id: string;
  kind: string;
  manifest?: SubtaskArtifactManifest;
}

export interface SessionStateView {
  session_id: string;
  trace_id: string;
  state: SessionState;
  created_at: string;
  pending_gates: readonly PendingGateView[];
  pending_clarification?: PendingClarificationView | null;
  subtasks: readonly SubtaskView[];
  latest_route_result?: NaturalLanguageRouteView | null;
  latest_intake_choice?: IntakeChoiceResponse | null;
  latest_intake_next_step?: IntakeNextStepView | null;
}

export const TERMINAL_STATES: ReadonlySet<SessionState> = new Set<SessionState>([
  "completed",
  "failed",
]);

/**
 * Phase 6 contract: `/chat` is only accepted in the `created` state.
 * Once the planner starts (`planning`), is awaiting approval, running,
 * or terminal, a new `/chat` would be rejected server-side (409) or
 * would overwrite the active plan. The UI mirrors this exactly so the
 * user can never type into a disallowed state.
 *
 * Note: `null` (no session yet) is also locked here; the store selector
 * surfaces "create a session first" separately.
 */
export function isChatLockedState(state: SessionState | null): boolean {
  return state !== "created";
}

File: apps/orchestrator/schemas.py
"""Public API schemas for the orchestrator (DESIGN.md §4.2).

The public API is the only stable contract — internal modules (planner,
cad generators, validator, slicer) keep evolving but these models pin
the wire format UI/CLI clients depend on. Treat any field change here
as a breaking change.
"""

from datetime import datetime
from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class SessionState(StrEnum):
    """Lifecycle states emitted by the orchestrator.

    States transition forward only; a terminal state (``COMPLETED`` /
    ``FAILED``) never returns to ``RUNNING``. UIs can render the value
    directly.
    """

    CREATED = "created"
    PLANNING = "planning"
    AWAITING_USER_INPUT = "awaiting_user_input"
    AWAITING_APPROVAL = "awaiting_approval"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class ApprovalDecision(StrEnum):
    """Outcome of a human approval gate (DESIGN.md §P9 — human-in-the-loop)."""

    APPROVED = "approved"
    REJECTED = "rejected"


class ApprovalStatus(StrEnum):
    """Lifecycle of a single ``ApprovalGate``."""

    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"


class ClarificationStatus(StrEnum):
    """Lifecycle of one ask-user clarification gate."""

    PENDING = "pending"
    ANSWERED = "answered"
    CANCELLED = "cancelled"
    EXPIRED = "expired"


class ClarificationAnswerType(StrEnum):
    """Small input vocabulary for clarification answers."""

    TEXT = "text"
    NUMBER = "number"
    CHOICE = "choice"


class IntakeChoiceAction(StrEnum):
    """Explicit user action for an asset intake decision.

    Phase 12V deliberately records a user choice without mutating the runtime
    asset catalog or starting CAD/STL/Orca execution.
    """

    USE_EXISTING_RUNTIME_ASSET = "use_existing_runtime_asset"
    REVIEW_EXISTING_DRAFT = "review_existing_draft"
    ADAPT_SIMILAR_CANDIDATE = "adapt_similar_candidate"
    CREATE_NEW_DRAFT = "create_new_draft"
    ASK_FOR_MORE_INFO = "ask_for_more_info"


class IntakeNextStepStatus(StrEnum):
    """Safe boundary opened after an explicit intake choice."""

    RUNTIME_ASSET_SELECTED = "runtime_asset_selected"
    DRAFT_REVIEW_HANDOFF = "draft_review_handoff"
    SIMILAR_CANDIDATE_REVIEW = "similar_candidate_review"
    DRAFT_AUTHORING_STARTED = "draft_authoring_started"
    INFORMATION_NEEDED = "information_needed"


class SessionCreateRequest(BaseModel):
    """Body for ``POST /session``."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    client_label: str | None = Field(default=None, max_length=120)


class SessionCreateResponse(BaseModel):
    """Response for ``POST /session``."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    trace_id: str = Field(min_length=1)
    state: SessionState
    created_at: datetime


class ChatRequest(BaseModel):
    """Body for ``POST /chat``."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    message: str = Field(min_length=1, max_length=4000)


class PendingGateView(BaseModel):
    """Approval gate visible to the client."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    gate_id: str = Field(min_length=1)
    prompt: str = Field(min_length=1)
    requested_at: datetime
    status: ApprovalStatus


class ClarificationQuestion(BaseModel):
    """A user-facing question produced by the decision layer."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str = Field(min_length=1)
    field: str = Field(min_length=1)
    question: str = Field(min_length=1)
    required: bool = True
    answer_type: ClarificationAnswerType = ClarificationAnswerType.TEXT
    choices: tuple[str, ...] = ()
    unit: str | None = Field(default=None, min_length=1)

    @model_validator(mode="after")
    def _choice_questions_need_choices(self) -> "ClarificationQuestion":
        if self.answer_type is ClarificationAnswerType.CHOICE and not self.choices:
            raise ValueError("choice clarification questions require choices.")
        return self


class PendingClarificationView(BaseModel):
    """Clarification gate visible to the client."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    clarification_id: str = Field(min_length=1)
    session_id: str = Field(min_length=1)
    subtask_id: str | None = Field(default=None, min_length=1)
    trace_id: str | None = Field(default=None, min_length=1)
    reason: str = Field(min_length=1)
    questions: tuple[ClarificationQuestion, ...] = Field(min_length=1, max_length=3)
    source_decision: dict[str, Any] = Field(default_factory=dict)
    plan_snapshot_id: str | None = Field(default=None, min_length=1)
    requirement_snapshot_id: str | None = Field(default=None, min_length=1)
    requested_at: datetime
    status: ClarificationStatus


class SubtaskView(BaseModel):
    """Planner-emitted subtask projected for the client."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str = Field(min_length=1)
    kind: Literal["mechanical", "organic"]
    description: str = Field(min_length=1)
    depends_on: tuple[str, ...] = ()


class AssetIntakeDecisionView(BaseModel):
    """User-facing asset intake decision before draft authoring."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    status: str = Field(min_length=1)
    reason: str = Field(min_length=1)
    runtime_asset_id: str | None = Field(default=None, min_length=1)
    draft_asset_id: str | None = Field(default=None, min_length=1)
    candidate_asset_ids: tuple[str, ...] = ()
    reference_candidate_ids: tuple[str, ...] = ()
    new_draft_allowed: bool
    metadata: dict[str, Any] = Field(default_factory=dict)


class LLMAssistShadowView(BaseModel):
    """Advisory local LLM shadow candidate exposed for comparison only.

    This view never changes product behavior. ``used_for_decision`` is always
    ``False`` and ``deterministic_route`` mirrors the actual product route.
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    mode: str = Field(min_length=1)
    succeeded: bool
    used_for_decision: bool
    model_name: str = Field(min_length=1)
    provider_name: str = Field(min_length=1)
    prompt_hash: str = Field(min_length=1)
    schema_version: str = Field(min_length=1)
    deterministic_route: str = Field(min_length=1)
    confidence: float | None = None
    fallback_reason: str | None = None
    candidate_route: str | None = None
    candidate_subjects: tuple[str, ...] = ()
    candidate_style: str | None = None
    clarification_question: str | None = None
    requirement_candidate: dict[str, Any] | None = None
    search_expansion: dict[str, Any] | None = None
    raw_response_excerpt: str | None = None


class NaturalLanguageRouteView(BaseModel):
    """Beginner prompt route decision exposed before runtime generation."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    selected_route: str = Field(min_length=1)
    route_snapshot_id: str = Field(min_length=1)
    reason: str = Field(min_length=1)
    request_id: str | None = Field(default=None, min_length=1)
    runtime_asset_id: str | None = Field(default=None, min_length=1)
    draft_asset_id: str | None = Field(default=None, min_length=1)
    candidate_asset_ids: tuple[str, ...] = ()
    clarification_required: bool
    clarification_questions: tuple[str, ...] = ()
    generation_allowed: bool
    requirement: dict[str, Any] = Field(default_factory=dict)
    intake_decision: AssetIntakeDecisionView | None = None
    shadow_result: LLMAssistShadowView | None = None


class IntakeChoiceRequest(BaseModel):
    """Body for ``POST /intake-choice``."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    action: IntakeChoiceAction
    route_snapshot_id: str = Field(min_length=1)


class IntakeChoiceResponse(BaseModel):
    """Response for ``POST /intake-choice``.

    The response is intentionally audit-heavy: it tells the UI what was
    accepted while preserving the Phase 12 boundary that no product/release
    approval and no runtime execution happened merely because the choice was
    recorded.
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    accepted: bool
    action: IntakeChoiceAction
    state: SessionState
    route_snapshot_id: str = Field(min_length=1)
    intake_status: str | None = None
    selected_route: str | None = None
    runtime_asset_id: str | None = None
    draft_asset_id: str | None = None
    candidate_asset_ids: tuple[str, ...] = ()
    reference_candidate_ids: tuple[str, ...] = ()
    draft_authoring_requested: bool = False
    runtime_execution_started: bool = False
    runtime_catalog_registered: bool = False
    product_ready: bool = False
    release_allowed: bool = False
    message: str = Field(min_length=1)


class IntakeNextStepRequest(BaseModel):
    """Body for ``POST /intake-next-step``."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    route_snapshot_id: str = Field(min_length=1)


class IntakeNextStepView(BaseModel):
    """Auditable next boundary derived from a recorded intake choice."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    action: IntakeChoiceAction
    status: IntakeNextStepStatus
    state: SessionState
    route_snapshot_id: str = Field(min_length=1)
    selected_route: str | None = None
    runtime_asset_id: str | None = None
    draft_asset_id: str | None = None
    candidate_asset_ids: tuple[str, ...] = ()
    reference_candidate_ids: tuple[str, ...] = ()
    draft_asset: dict[str, Any] | None = None
    draft_authoring_status: str | None = None
    draft_authoring_reason: str | None = None
    # fallback ADR item 6: structured IP/abuse pre-filter decision (decision /
    # matched_rule_id / reason_code / user_message_ko) when the capable model was
    # blocked BEFORE being invoked. Optional additive field: existing responses
    # now serialize `"prefilter": null`.
    prefilter: dict[str, Any] | None = None
    review_required: bool = True
    runtime_execution_started: bool = False
    runtime_catalog_registered: bool = False
    product_ready: bool = False
    release_allowed: bool = False
    message: str = Field(min_length=1)


class RuntimeAssetExecutionRequest(BaseModel):
    """Body for ``POST /runtime-asset-execution``.

    This starts manufacturing-runtime verification only after a prior
    ``runtime_asset_selected`` next-step boundary. Choosing an intake action
    alone never starts CAD/STL/Orca execution.
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    route_snapshot_id: str = Field(min_length=1)


class RuntimeAssetExecutionResponse(BaseModel):
    """Result of explicitly executing a selected curated runtime asset."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    state: SessionState
    route_snapshot_id: str = Field(min_length=1)
    runtime_asset_id: str = Field(min_length=1)
    subtask_id: str = Field(min_length=1)
    runtime_execution_started: bool = True
    runtime_catalog_registered: bool = False
    product_ready: bool = False
    release_allowed: bool = False
    next_step: IntakeNextStepView
    manifest: dict[str, Any]
    message: str = Field(min_length=1)


class ChatResponse(BaseModel):
    """Response for ``POST /chat``.

    Exactly one of ``subtasks`` / ``clarifying_questions`` is populated
    when the planner returns successfully. ``state`` is the post-call
    session state so the client does not need a follow-up GET.

    A pre-filter block is a *fourth* shape: every planner output field stays
    empty and ``prefilter`` carries the verdict. That combination is already
    legal — the validator below only rejects more than one populated field, so
    zero needs no exemption.
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    trace_id: str = Field(min_length=1)
    state: SessionState
    subtasks: tuple[SubtaskView, ...] = ()
    clarifying_questions: tuple[str, ...] = ()
    pending_gate: PendingGateView | None = None
    pending_clarification: PendingClarificationView | None = None
    route_result: NaturalLanguageRouteView | None = None
    # fallback ADR item 6: structured IP/abuse pre-filter decision (decision /
    # matched_rule_id / reason_code / user_message_ko) when the raw ``/chat``
    # message was blocked BEFORE the planner was invoked. Mirrors
    # ``IntakeNextStepView.prefilter`` so both blocked boundaries carry the
    # verdict in the same shape. Optional additive field: existing responses
    # now serialize ``"prefilter": null``.
    #
    # Deliberately NOT folded into ``clarifying_questions``: "we need more
    # information" and "we do not accept this request" are different events,
    # and a client that conflates them shows a re-ask UI for a refusal.
    prefilter: dict[str, Any] | None = None

    @model_validator(mode="after")
    def _exactly_one_planner_output(self) -> "ChatResponse":
        populated = sum(
            1
            for has_value in (
                bool(self.subtasks),
                bool(self.clarifying_questions),
                self.pending_clarification is not None,
            )
            if has_value
        )
        if populated > 1:
            raise ValueError(
                "ChatResponse may not carry subtasks, clarifying_questions, and "
                "pending_clarification in the same reply; planner must commit "
                "to one."
            )
        return self


class ApproveRequest(BaseModel):
    """Body for ``POST /approve``."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    gate_id: str = Field(min_length=1)
    decision: ApprovalDecision
    comments: str | None = Field(default=None, max_length=2000)


class ApproveJobRequest(BaseModel):
    """Body for ``POST /approve/{job_id}``.

    Phase 6 treats ``job_id`` as the approval gate id. A later durable job
    model can map the path id to the gate without changing this public route.
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    decision: ApprovalDecision
    comments: str | None = Field(default=None, max_length=2000)


class ApproveResponse(BaseModel):
    """Response for ``POST /approve``."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    gate_id: str = Field(min_length=1)
    state: SessionState
    decision: ApprovalDecision


ClarificationAnswerValue = str | int | float


class ClarifyRequest(BaseModel):
    """Body for ``POST /clarify``."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    clarification_id: str = Field(min_length=1)
    answers: dict[str, ClarificationAnswerValue] = Field(min_length=1)


class ClarifyResponse(BaseModel):
    """Response for ``POST /clarify``."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    clarification_id: str = Field(min_length=1)
    accepted: bool
    state: SessionState
    route_result: NaturalLanguageRouteView | None = None
    pending_clarification: PendingClarificationView | None = None


class SessionEventKind(StrEnum):
    """Discriminator for ``SessionEvent`` payloads streamed over WS."""

    STATE_CHANGED = "state_changed"
    NATURAL_LANGUAGE_ROUTE_SELECTED = "natural_language_route_selected"
    INTAKE_CHOICE_RECORDED = "intake_choice_recorded"
    INTAKE_NEXT_STEP_RECORDED = "intake_next_step_recorded"
    PLANNING_COMPLETED = "planning_completed"
    SUBTASK_STARTED = "subtask_started"
    SUBTASK_PROGRESS = "subtask_progress"
    SUBTASK_COMPLETED = "subtask_completed"
    SUBTASK_FAILED = "subtask_failed"
    APPROVAL_REQUESTED = "approval_requested"
    APPROVAL_RESOLVED = "approval_resolved"
    CLARIFICATION_REQUESTED = "clarification_requested"
    CLARIFICATION_RESOLVED = "clarification_resolved"
    ERROR = "error"


class SessionEvent(BaseModel):
    """Server-pushed event delivered through ``WS /ws/{session_id}``.

    Payload schema is intentionally open (``dict[str, Any]``); UIs read
    the discriminator first, then the typed payload. Keeping payload
    open lets internal events evolve without rev-bumping the public WS
    contract.
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    trace_id: str = Field(min_length=1)
    kind: SessionEventKind
    state: SessionState
    occurred_at: datetime
    payload: dict[str, Any] = Field(default_factory=dict)


class SessionStateView(BaseModel):
    """Full session snapshot returned by ``GET /session/{id}``."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    session_id: str = Field(min_length=1)
    trace_id: str = Field(min_length=1)
    state: SessionState
    created_at: datetime
    pending_gates: tuple[PendingGateView, ...] = ()
    pending_clarification: PendingClarificationView | None = None
    subtasks: tuple[SubtaskView, ...] = ()
    latest_route_result: NaturalLanguageRouteView | None = None
    latest_intake_choice: IntakeChoiceResponse | None = None
    latest_intake_next_step: IntakeNextStepView | None = None


__all__ = [
    "ApprovalDecision",
    "ApprovalStatus",
    "ApproveJobRequest",
    "ApproveRequest",
    "ApproveResponse",
    "ClarificationAnswerType",
    "ClarificationAnswerValue",
    "ClarificationQuestion",
    "ClarificationStatus",
    "ClarifyRequest",
    "ClarifyResponse",
    "ChatRequest",
    "ChatResponse",
    "AssetIntakeDecisionView",
    "IntakeChoiceAction",
    "IntakeChoiceRequest",
    "IntakeChoiceResponse",
    "IntakeNextStepRequest",
    "IntakeNextStepStatus",
    "IntakeNextStepView",
    "NaturalLanguageRouteView",
    "PendingGateView",
    "PendingClarificationView",
    "RuntimeAssetExecutionRequest",
    "RuntimeAssetExecutionResponse",
    "SessionCreateRequest",
    "SessionCreateResponse",
    "SessionEvent",
    "SessionEventKind",
    "SessionState",
    "SessionStateView",
    "SubtaskView",
]
